3/3/2026 
## NUMPY 

In [3]:
import numpy as np
print("Successfully Installed")

Successfully Installed


In [4]:
arr=np.array([1,2,3,4])
arr 

array([1, 2, 3, 4])

In [16]:
arr=np.array([[1,2,3],[4,5,6]])

print("Shape(rows and columns):", arr.shape)
print("Dimension:",arr.ndim)
print("Size:", arr.size)
print("Data-Type:", arr.dtype)

Shape(rows and columns): (2, 3)
Dimension: 2
Size: 6
Data-Type: int64


In [15]:
#MATHEMATICAL OPERATIONS
a=np.array([1,2,3])
b=np.array([4,5,6])

print("Addition:", a+b)
print("Subtraction:", a-b)
print("Multiplication:", a*b)
print("Power:", a**b)
print("Division:", a/b)

Addition: [5 7 9]
Subtraction: [-3 -3 -3]
Multiplication: [ 4 10 18]
Power: [  1  32 729]
Division: [0.25 0.4  0.5 ]


In [34]:
## 
arr=np.array([1,2,3,4,5])

print(arr)
print("Sum:", np.sum(arr))
print("Mean:", np.mean(arr))
print("Maximum:", np.max(arr))
print("Minimum:", np.min(arr))
print("Standard Deviation:", np.std(arr))
print("Variance:", np.var(arr))





[1 2 3 4 5]
Sum: 15
Mean: 3.0
Maximum: 5
Minimum: 1
Standard Deviation: 1.4142135623730951
Variance: 2.0


In [43]:
## BOOLEAN INDEXING
Scores=np.array([75,80,40,90])
Scores
#Scores[Scores >= 70]
#Scores[Scores <=70]
#Scores[Scores <40]


array([75, 80, 40, 90])

In [94]:
#Select the rows where the first column is greater than 30

Matrix=np.array([[10,20,30],
               [40,50,60],
               [70,80,90]])

Matrix
filtered_rows=Matrix[Matrix[:,0]>30] #":" means start from the beginning, "0" the first column (axes)
filtered_rows

array([[40, 50, 60],
       [70, 80, 90]])

In [79]:
#Select the rows where the last column is greater than 60
Matrix=np.array([[10,50,30],
               [40,50,60],
               [70,80,90]])
filtered_rows=Matrix[Matrix[:,2]>60] #":" means start from the beginning, "2" the last column (axes)
filtered_rows


array([[70, 80, 90]])

In [111]:
#Select the row where the second column is less than 80
Matrix=np.array([[10,50,30],
               [40,50,60],
               [70,80,90]])
filtered_rows=Matrix[Matrix[:,1]<80]  #":" means start from the beginning (Row), "2" the last column (Column)
filtered_rows


array([[10, 50, 30],
       [40, 50, 60]])

#Select the column where the last row is greater than 60
Matrix=np.array([[10,50,30],
               [40,50,60],
               [70,80,90]])
filtered_rows=Matrix[Matrix[2,:]>60] #"2" means last row, ": the entire column (axes)
filtered_rows


In [117]:
Matrix=np.array([[10,50,30],
               [40,50,60],
               [70,80,90]])
number=50
filtered_rows= np.all(Matrix < number, axis=1)
print(filtered_rows)


[False False False]


In [100]:
arr=np.array([10,20,30,40, 50])
filtered=arr[(arr > 20) & (arr < 50)]
condition=arr[(arr <10)|(arr > 40)]

print(filtered)
print(condition)

[30 40]
[50]


ASSIGNMENT
Students_Scores=np.array([[65,70,80],
                        [45,55,60],
                        [90,95,85],
                        [70,60,75],
                        [50,40,65]])
select all rows where the first column (Test1)>60
select all rows where the first column (Test1)<=60
select  rows where  Test1>60 and Test3>=75
select  rows where  Test1<70 and Test2<60
Replace all scores in Test3 that are less than 70 with 70

In [121]:
import numpy as np

Students_Scores = np.array([[65,70,80],
                            [45,55,60],
                            [90,95,85],
                            [70,60,75],
                            [50,40,65]])

# 1. Select all rows where Test1 > 60
rows_test1_gt_60 = Students_Scores[Students_Scores[:,0] > 60]
print("Test1 > 60:\n", rows_test1_gt_60)

# 2. Select all rows where Test1 <= 60
rows_test1_le_60 = Students_Scores[Students_Scores[:,0] <= 60]
print("Test1 <= 60:\n", rows_test1_le_60)

# 3. Select rows where Test1 > 60 AND Test3 >= 75
rows_condition1 = Students_Scores[(Students_Scores[:,0] > 60) & 
                                   (Students_Scores[:,2] >= 75)]
print("Test1 > 60 and Test3 >= 75:\n", rows_condition1)

# 4. Select rows where Test1 < 70 AND Test2 < 60
rows_condition2 = Students_Scores[(Students_Scores[:,0] < 70) & 
                                   (Students_Scores[:,1] < 60)]
print("Test1 < 70 and Test2 < 60:\n", rows_condition2)

# 5. Replace all scores in Test3 that are less than 70 with 70
Students_Scores[Students_Scores[:,2] < 70, 2] = 70
print("Updated Students_Scores:\n", Students_Scores)

Test1 > 60:
 [[65 70 80]
 [90 95 85]
 [70 60 75]]
Test1 <= 60:
 [[45 55 60]
 [50 40 65]]
Test1 > 60 and Test3 >= 75:
 [[65 70 80]
 [90 95 85]
 [70 60 75]]
Test1 < 70 and Test2 < 60:
 [[45 55 60]
 [50 40 65]]
Updated Students_Scores:
 [[65 70 80]
 [45 55 70]
 [90 95 85]
 [70 60 75]
 [50 40 70]]


## 4/3/2026   CLASS WORK
A teacher has exam scores for 5 students in 3 subjects
1. Maths
2. Science
3. English
4. Each row represent a student
5. Each column represent each subject

   Scores=np.array([[85, 78, 92],
                   [70, 88, 75],
                   [90, 95, 93],
                   [60, 65, 70],
                   [88, 84, 91]
                   ])
a. What is the average score of each student?
b. What is the average score per subject?
c. Find the top performance
d. Find the highest score in each subject
e. Find the Students who passed all subjects, assuming pass mark is 70


In [2]:
import numpy as np
Scores = np.array([[85, 78, 92],
                   [70, 88, 75],
                   [90, 95, 93],
                   [60, 65, 70],
                   [88, 84, 91]
                   ])
print("Scores:\n", Scores)

#Average score of each student
Each_Student_Average = np.mean(Scores, axis=1)
Each_Student_Average = np.round(Each_Student_Average,0)
print("Each Student's Average:\n",Each_Student_Average)

#Average score per subject
Each_Subject_Average = np.mean(Scores, axis=0)
Each_Student_Average = np.round(Each_Subject_Average,0)
print("Average of Each Student:\n",Each_Subject_Average)

#Top_Performance
Best_Performance = np.argmax(Scores, axis=0)
print("Best Performance per subject:\n",Best_Performance)

#Best_Subject= [Scores, axis=0]
#Top_Student=np.argmax(Best_Subject)
#print("Total Scores:\n",Best_Subject)
#print("Total Scores:\n",Top_Student)

Total_Scores=np.sum(Scores, axis=1)
Top_Student=np.argmax(Total_Scores)
print("Total Scores:\n",Total_Scores)
print("Total Scores:\n",Top_Student)


Student_Best_Subject = np.argmax(Scores, axis=1)
print("Student Best Subject:\n",Student_Best_Subject)

#Highest score
Highest_Score = np.max(Scores, axis=0)
print("Maximum Score:\n",Highest_Score)

#the Students who passed all subjects, assuming a pass mark is 70
Passed=Scores>=70
print("Students who passed all the subjects:\n",Passed)

Student_Passed_All_Subects= Scores[(Scores[:,0] >=70)]  
print("Student who passed all subjects:\n",Student_Passed_All_Subects)

Good_Performance=np.all(Scores >=70, axis=1)
print("Good Performance:\n", Good_Performance)



Scores:
 [[85 78 92]
 [70 88 75]
 [90 95 93]
 [60 65 70]
 [88 84 91]]
Each Student's Average:
 [85. 78. 93. 65. 88.]
Average of Each Student:
 [78.6 82.  84.2]
Best Performance per subject:
 [2 2 2]
Total Scores:
 [255 233 278 195 263]
Total Scores:
 2
Student Best Subject:
 [2 1 1 2 2]
Maximum Score:
 [90 95 93]
Students who passed all the subjects:
 [[ True  True  True]
 [ True  True  True]
 [ True  True  True]
 [False False  True]
 [ True  True  True]]
Student who passed all subjects:
 [[85 78 92]
 [70 88 75]
 [90 95 93]
 [88 84 91]]
Good Performance:
 [ True  True  True False  True]


a: The Best Student is student 3 in column 2
b: The hardest subject, according to their performance, is Mathematics
c: The number of students who failed a subject based on the benchmark of pass given is one, Student 4 in column 3

In Summary

## 6/03/2026-
## QUESTION
Supermarket Sales Analysis
A small supermarket sells 4 products:
a. Bread
b. Milk
c. Eggs
d. Rice

The store recorded the number of items  and sold over 7 days (one week)
a. Each row= one day 
b. Each column=one product

1. Check the shape of the array
2. Calculate the total items sold each day
3. Calculate the total sales for each product during the week
4. Calculate the average daily sales for each product (Rounding up to 2dp)
5. Which product sold the most during the week?
6. Which day has the highest total sales?
7. Which product had the lowest average sales

In [37]:
#rows = days (Mon-Sun)
#column= Bread, milk, eggs, and rice
Sales = np.array([
    [50, 30, 40, 20],   
    [60, 35, 45, 25],   
    [55, 32, 50, 22],   
    [70, 40, 60, 30],   
    [65, 38, 55, 28],   
    [80, 45, 65, 35],   
    [75, 42, 62, 33]    
])

Sales
#Shape of the Array
print("Shape of the Array:", Sales.shape)

#Total Items sold each day
Total_Sales=np.sum(Sales, axis=1)
print("Total Daily Sales for each Day:", Total_Sales)

#Total Sales for each product during the week
Total_Sales_Week=np.sum(Sales, axis=0)
print("Total Weekly Sales for each product:", Total_Sales_Week)

#Average Daily Sales for each  product 
Average_Sales=np.mean(Sales, axis=0)
Average_Sales=np.round(Average_Sales, 2)
print("Average Sales for Each Day:",Average_Sales )

#Most Sold Product per week
Most_Sold_Product=np.max(Sales, axis=0)
print("Most Sold Product Per Week:",Most_Sold_Product )

#The Day with the Highest Total Sales
Most_Sold_Product_per_Day =np.max(Sales, axis=1)
Total_Sales_of_Most_Sold_Product_Day= np.sum(Most_Sold_Product_per_Day, axis=0)
print("Most_Sold_Product_per_Day:",Most_Sold_Product_per_Day, "Bread")
print("Total_Sales_of_Most_Sold_Product_Day (Bread):",Total_Sales_of_Most_Sold_Product_Day )

#The Day with the lowest total Sales
Least_Sold_Product_per_Day =np.min(Sales, axis=1)
Total_Sales_of_least_Sold_Product_Day= np.mean (Most_Sold_Product_per_Day, axis=0)
print("least Sold Product per Day:",Least_Sold_Product_per_Day)
print("Average of least Sold Product Day",Total_Sales_of_least_Sold_Product_Day)


#REMARK


Shape of the Array: (7, 4)
Total Daily Sales for each Day: [140 165 159 200 186 225 212]
Total Weekly Sales for each product: [455 262 377 193]
Average Sales for Each Day: [65.   37.43 53.86 27.57]
Most Sold Product Per Week: [80 45 65 35]
Most_Sold_Product_per_Day: [50 60 55 70 65 80 75] Bread
Total_Sales_of_Most_Sold_Product_Day (Bread): 455
least Sold Product per Day: [20 25 22 30 28 35 33]
Average of least Sold Product Day 65.0


## AS A DATA ANALYST
Student Performance Analytics
You are a data analyst for a University. The University wants to analyse student performance for one week of assessments across multiple subjects. The goal is to identify top students and overall class performance using numpy.

Dataset
The class has 8 students and 4 subjects:
- Mathematics
- English
- Science
- Computer
Each row represents a student, and each column represents a subject score
1. Find the total score for each student
2. Calculate the average score for each student
3. Determine the top-performing Student
4. Calculate the lowest performing student
5. Identify the most difficult subject
6. Identify the best-performing subject
7. Find the highest score in the dataset
8. Find the lowest score in the dataset
9. Calculate the overall class average score
10. Assuming 70 and above = Pass and below 70=Fail. Determine which students passed all subjects
11. Determine which student failed at least one subject
12. Find how many students score above 80 in computer
13. Determine which student scored the highest in Mathematics
14. Display the student who scored above the class average


Students = np.array([
"Moyo", "Charity", "Honour", "Ifeoluwa",
"Treasure", "Itunuoluwa", "David", "Grace"
])

Subjects = np.array([
"Mathematics", "English", "Agric", "Computer"
])

Scores = np.array([
[78, 65, 80, 90],
[85, 70, 75, 88],
[60, 75, 70, 72],
[90, 88, 92, 95],
[55, 60, 58, 65],
[72, 68, 74, 80],
[88, 82, 85, 90],
[70, 72, 78, 76]
])


Insight
- Which Student performed Best Overall
- Which subject do students struggle with the most
- How the school could improve performance
- Whether the class average is good or poor

In [4]:
import numpy as np
Students = np.array([
"Moyo", "Charity", "Honour", "Ifeoluwa",
"Treasure", "Itunuoluwa", "David", "Grace"
])

Subjects = np.array([
"Mathematics", "English", "Agric", "Computer"
])

Scores = np.array([
[78, 65, 80, 90],
[85, 70, 75, 88],
[60, 75, 70, 72],
[90, 88, 92, 95],
[55, 60, 58, 65],
[72, 68, 74, 80],
[88, 82, 85, 90],
[70, 72, 78, 76]
])

print("Students:\n", Students)
print("Subjects:\n", Subjects)
print("Scores:\n", Scores)

Shape_Array=np.shape(Scores)
print("Shape:", Shape_Array)

import numpy as np

# Student names
Students = np.array([
"Moyo", "Charity", "Honour", "Ifeoluwa",
"Treasure", "Itunuoluwa", "David", "Grace"
])

# Subjects
Subjects = np.array([
"Mathematics", "English", "Agric", "Computer"
])

# Scores dataset
Scores = np.array([
[78, 65, 80, 90],
[85, 70, 75, 88],
[60, 75, 70, 72],
[90, 88, 92, 95],
[55, 60, 58, 65],
[72, 68, 74, 80],
[88, 82, 85, 90],
[70, 72, 78, 76]
])

# 1. Total score for each student
total_scores = Scores.sum(axis=1)
print("Total Scores:", total_scores)

# 2. Average score for each student
average_scores = Scores.mean(axis=1)
print("Average Scores:", average_scores)

# 3. Top performing student
top_student_index = np.argmax(total_scores)
top_student = Students[top_student_index]
print("Top Student:", top_student)

# 4. Lowest performing student
lowest_student_index = np.argmin(total_scores)
lowest_student = Students[lowest_student_index]
print("Lowest Student:", lowest_student)

# 5. Most difficult subject (lowest average)
subject_avg = Scores.mean(axis=0)
difficult_subject = Subjects[np.argmin(subject_avg)]
print("Most Difficult Subject:", difficult_subject)

# 6. Best performing subject (highest average)
best_subject = Subjects[np.argmax(subject_avg)]
print("Best Performing Subject:", best_subject)

# 7. Highest score in dataset
highest_score = Scores.max()
print("Highest Score:", highest_score)

# 8. Lowest score in the dataset
lowest_score = Scores.min()
print("Lowest Score:", lowest_score)

# 9. Overall class average
class_average = Scores.mean()
print("Class Average:", class_average)

# 10. Students who passed all subjects (>=70)
passed_all = Students[np.all(Scores >= 70, axis=1)]

print("Students who passed all subjects:", passed_all)

# 11. Students who failed at least one subject
failed_one = Students[np.any(Scores < 70, axis=1)]
print("Students who failed at least one subject:", failed_one)

# 12. Students scoring above 80 in Computer
computer_above_80 = Students[Scores[:,3] > 80]
print("Students scoring above 80 in Computer:", computer_above_80)

# 13. Student with highest Mathematics score
math_top = Students[np.argmax(Scores[:,0])]
print("Top Mathematics Student:", math_top)

# 14. Students above class average
above_avg_students = Students[average_scores > class_average]
print("Students above class average:", above_avg_students)


Students:
 ['Moyo' 'Charity' 'Honour' 'Ifeoluwa' 'Treasure' 'Itunuoluwa' 'David'
 'Grace']
Subjects:
 ['Mathematics' 'English' 'Agric' 'Computer']
Scores:
 [[78 65 80 90]
 [85 70 75 88]
 [60 75 70 72]
 [90 88 92 95]
 [55 60 58 65]
 [72 68 74 80]
 [88 82 85 90]
 [70 72 78 76]]
Shape: (8, 4)
Total Scores: [313 318 277 365 238 294 345 296]
Average Scores: [78.25 79.5  69.25 91.25 59.5  73.5  86.25 74.  ]
Top Student: Ifeoluwa
Lowest Student: Treasure
Most Difficult Subject: English
Best Performing Subject: Computer
Highest Score: 95
Lowest Score: 55
Class Average: 76.4375
Students who passed all subjects: ['Charity' 'Ifeoluwa' 'David' 'Grace']
Students who failed at least one subject: ['Moyo' 'Honour' 'Treasure' 'Itunuoluwa']
Students scoring above 80 in Computer: ['Moyo' 'Charity' 'Ifeoluwa' 'David']
Top Mathematics Student: Ifeoluwa
Students above class average: ['Moyo' 'Charity' 'Ifeoluwa' 'David']


## INSIGHT
From the analysis of the dataset, the Overall Best student had an average of 365, and the subject the students performed best in is Computer.
The lowest average performance is 238, and the subject in which the students performed poorly is English. The overall average performance of the class is okay but more attention should be given to the students who are below the passmarks. 
The School can improve on their teaching methods, then also call the students guidance and discuss with them on how to improve the performace of their children because it is not the work of the teachers alone but also on the students and guidance. 
There should be extra-moral classes for the students, especially on the subject they performed poorly. They should also check the method of the Teacher's way of teaching

10/03/2026
How to General Random Numbers and Round up

In [20]:
import numpy as np
#Generate a random numpy array
arr=np.random.random(5)
arr1=np.round(arr, 2)
print(arr)
print(arr1)

#Generate a random numpy array (Matrix form
arrr=np.random.random((3,5))
arr2=np.round(arrr, 2)
print("\n", arrr)
print(arr2)

#Generate a random numpy Integer
array=np.random.randint(1, 10, size=5)
print("\n", array)

#Generate a random numpy Matrix
array=np.random.randint(1, 100, size=(5,5))
print("\n", array)



[0.87982222 0.17439714 0.56637756 0.18894969 0.76745304]
[0.88 0.17 0.57 0.19 0.77]

 [[0.95326084 0.11583754 0.01406039 0.07566415 0.77810995]
 [0.86647018 0.61190513 0.65105048 0.67368068 0.81058537]
 [0.45723193 0.38721228 0.13487199 0.92590479 0.68976233]]
[[0.95 0.12 0.01 0.08 0.78]
 [0.87 0.61 0.65 0.67 0.81]
 [0.46 0.39 0.13 0.93 0.69]]

 [5 5 2 3 8]

 [[85 35 85 30 40]
 [10 77 58 52 47]
 [61  4 90  8  1]
 [89 76 38 31  6]
 [19 10 90 57 87]]


## IF YOU ARE GIVEN ANY WORK, ENSURE TO GIVE AN INSIGHT ABOUT THE ANALYSIS ----- SHORT NOTE(LIKE REPORT AND SUGGESTION)